Custom metrics for Media

smatei opened the pull request:
@rviscomi, here is another custom metric for the [2020 media chapter](https://github.com/HTTPArchive/almanac.httparchive.org/issues/900#issuecomment-660275165). 

[source/img with sizes attribute expensive sql with regexp](https://github.com/HTTPArchive/almanac.httparchive.org/blob/2189e7b3d80b94a09c9a1c59926ea63b3fe17c34/sql/2019/04_Media/04_06b.sql)

I combined both conditions in the same metric, just as in the original sql. They could also be split metrics. Also we could add an empty attribute condition - img[sizes]:not([sizes=""]). Please advise.

As for the tests:

[img sizes](https://www.webpagetest.org/custom_metrics.php?test=200720_ZN_74bb4163bcdd2aa25c0e58aeca955708&run=1&cached=0)

[source sizes](https://www.webpagetest.org/custom_metrics.php?test=200720_YZ_5f276fed83e36382e40654e2d714787c&run=1&cached=0)

[no img/source sizes](https://www.webpagetest.org/custom_metrics.php?test=200720_41_fbe1329b555858d0bacaec7650a59f03&run=1&cached=0)

Tiggerito commented:
I'm in the SEO chapter and will be adding the gathering of img tag info. So far I'm checking alt, loading, width and height attributes. My plan is to place all the info in an 'image' property for the response. It could be worth combining this image based effort?

bazzadp commented:
`aria-label` and `aria-labelledby` attributes might also be interesting as alternatives to `alt` attributes.

smatei commented:
> 
> 
> I'm in the SEO chapter and will be adding the gathering of img tag info. So far I'm checking alt, loading, width and height attributes. My plan is to place all the info in an 'image' property for the response. It could be worth combining this image based effort?

@Tiggerito Yes, we could group them together under "image". But, I am new here, it is not my call. I will wait for further instructions from the code reviewers.

Tiggerito commented:
> @Tiggerito Yes, we could group them together under "image". But, I am new here, it is not my call. I will wait for further instructions from the code reviewers.

Don't worry, I'm also learning on the fly. Lets just keep in contact about what we need, and try to pull things together when we can.

You can see my images code here

https://github.com/Tiggerito/legacy.httparchive.org/blob/patch-1/custom_metrics/almanac.js

You used the term image not images for the property. You are probably right.

rviscomi commented on custom_metrics/almanac.js:
```diff
@@ -307,5 +307,7 @@ return JSON.stringify({
-  'has_picture_img': document.querySelectorAll('picture img').length > 0
+  'has_picture_img': document.querySelectorAll('picture img').length > 0,
+  // check if there are any source or img tags with sizes attribute
+  'has_image_sizes': document.querySelectorAll('source[sizes], img[sizes]').length > 0  
```
It might be more versatile to return the length itself rather than the `> 0` condition. This enables followup queries like "median number of images with sizes per page" for example.

(Also applicable to `has_picture_img`. You may also want to rename this from "has" to "num" or similar.)

smatei commented:
I have changed both custom metrics to return length, renamed them with num_ and updated tests for all combinations:

["num_picture_img":0,"num_image_sizes":0](https://www.webpagetest.org/custom_metrics.php?test=200721_EZ_25a3478579a05e6580ae866295f8dcce&run=1&cached=0)

["num_picture_img":0,"num_image_sizes":1](https://www.webpagetest.org/custom_metrics.php?test=200721_NV_35822e5309f38f32513d412d86bc0f76&run=1&cached=0)

["num_picture_img":1,"num_image_sizes":0](https://www.webpagetest.org/custom_metrics.php?test=200721_V6_6804133144bad9ab4bb82de32c8c9fa7&run=1&cached=0)

["num_picture_img":1,"num_image_sizes":2](https://www.webpagetest.org/custom_metrics.php?test=200721_ME_1caa380bd0da65431e8f0fce2842cfd4&run=1&cached=0)

Tiggerito commented:
I've added a [pull request](https://github.com/HTTPArchive/legacy.httparchive.org/pull/171/files) from the SEO team for our image property. I think it may be tricky to consolidate things at this point, and we are on a tight schedule. 

Maybe once my code is merged you can consider folding your data into the image property. That way you know if your own property has moved and can adjust your SQL to match. I'd say not urgent, as it's just about neatness.

I've also noticed quite a mix of property naming conventions, and am not sure which one is best. It's always confusing when we are coding in multiple languages that follow different rules. Again, not important with this tight schedule. Most important is that we get the data needed.

smatei commented:
I am taking my chance with

[Video player attributes](
https://github.com/HTTPArchive/almanac.httparchive.org/blob/2189e7b3d80b94a09c9a1c59926ea63b3fe17c34/sql/2019/04_Media/04_19.sql)

Tests:

["video-nodes-attributes":["controls","width"]](https://www.webpagetest.org/custom_metrics.php?test=200724_CR_10468e9f7ad160b21310334227b14ce8&run=1&cached=0)

["video-nodes-attributes":[]](https://www.webpagetest.org/custom_metrics.php?test=200724_HA_c7bd1a696cb2700cace24047c60a5475&run=1&cached=0)

smatei commented:
We need srcset w/wo sizes for [Media](https://docs.google.com/document/d/1ip2At9j6ba17odQwBbXDVU7qiLnHRnX8DflAfdQNb9c/edit#).

I have added some metrics:

["num_srcset_all":6,"num_srcset_sizes":0](https://www.webpagetest.org/custom_metrics.php?test=200725_D1_1a1468d26c4747fdc55ab855ca4bd6d0&run=1&cached=0)

["num_srcset_all":1,"num_srcset_sizes":1](https://www.webpagetest.org/custom_metrics.php?test=200725_GT_df94fcfe4cf4930a4d27b7c055ef1019&run=1&cached=0)

["num_srcset_all":0,"num_srcset_sizes":0](https://www.webpagetest.org/custom_metrics.php?test=200725_JX_e4b90ba7bdc1ec1ca5c36921cb24b23d&run=1&cached=0)

smatei commented:
We need srcset with x/w descriptors custom metrics in [Media](https://docs.google.com/document/d/1ip2At9j6ba17odQwBbXDVU7qiLnHRnX8DflAfdQNb9c/edit#)

Tests:

"num_srcset_descriptor_x":2,"num_srcset_descriptor_w":2
https://www.webpagetest.org/custom_metrics.php?test=200725_28_cdc067a2a1ac8e4ae772326e7847f24a&run=1&cached=0

num_srcset_descriptor_x":0,"num_srcset_descriptor_w":0
https://www.webpagetest.org/custom_metrics.php?test=200725_9K_420b55dfb54681f73a19a6f731b8b3f8&run=1&cached=0

smatei commented:
We need the number of scrset candidates for [Media](https://docs.google.com/document/d/1ip2At9j6ba17odQwBbXDVU7qiLnHRnX8DflAfdQNb9c/edit#)

Tests:

["num_srcset_candidates":2](https://www.webpagetest.org/custom_metrics.php?test=200725_9P_be588dec8667baf67077cb64140769f4&run=1&cached=0)

["num_srcset_candidates":0](https://www.webpagetest.org/custom_metrics.php?test=200725_DT_a6d9638dc118b7346622479ef9bed3e8&run=1&cached=0)

smatei commented:
We need picture format switching for [Media](https://docs.google.com/document/d/1ip2At9j6ba17odQwBbXDVU7qiLnHRnX8DflAfdQNb9c/edit#)

Tests

["picture_formats":["image/webp"]](https://www.webpagetest.org/custom_metrics.php?test=200725_DJ_2184fbe385d6652ef7373b50de521d72&run=1&cached=0)

["picture_formats":[]](https://www.webpagetest.org/custom_metrics.php?test=200725_6C_42a10fde0c185b6de2254fdac25c82b7&run=1&cached=0)

smatei commented:
We need to count the picture tags that use min-resolution in [Media](https://docs.google.com/document/d/1ip2At9j6ba17odQwBbXDVU7qiLnHRnX8DflAfdQNb9c/edit#)

Tests:

["num_picture_using_min_resolution":52](https://www.webpagetest.org/custom_metrics.php?test=200727_ER_ac92e6cf8e13f810bf63d147f9808fa4&run=1&cached=0)

["num_picture_using_min_resolution":0](https://www.webpagetest.org/custom_metrics.php?test=200727_Q0_9df44a4668f1f3415be4d3e0716b84b6&run=1&cached=0)

smatei commented:
I have added a metric for counting pictures with device orientation for [Media](url)

Tests:

I could not find a live url for which that metric would return non 0 values, please help.
 
For 0 value - ["num_picture_using_orientation":0](https://www.webpagetest.org/custom_metrics.php?test=200727_NT_fd8531c7dc976d55aad8948c10faf47c&run=1&cached=0)

smatei commented:
I have added a custom metric for counting video nodes in [Media](https://github.com/HTTPArchive/legacy.httparchive.org/pull/url) 

Tests:

["num_video_nodes":2](https://www.webpagetest.org/custom_metrics.php?test=200727_CC_72196b2ab7d36e36c9b571ed1e456391&run=1&cached=0)

["num_video_nodes":0](https://www.webpagetest.org/custom_metrics.php?test=200727_B7_8d7257a0444e06ce5266992587f78104&run=1&cached=0)

rviscomi commented on custom_metrics/almanac.js:
```diff
@@ -306,6 +336,57 @@ return JSON.stringify({
+  })(),
+  // Counts the number of pictures using source media min-resolution
+  'num_picture_using_min_resolution': (() => {
+    var nodes = document.querySelectorAll('picture source[media*="min-resolution"]');
+    return countUniqueParents(nodes);
```
May I suggest a simpler approach for this and `num_picture_using_orientation` below?

```suggestion
    var pictures = document.querySelectorAll('picture');
    return Array.from(pictures).filter(picture => {
      return picture.querySelector('source[media*="min-resolution"]');
    }).length;
```

rviscomi commented on custom_metrics/almanac.js:
```diff
@@ -41,6 +41,36 @@ function parseNodes(nodes) {
 
+// Return the set of attributes for nodes,
+function getNodesAttributes(nodes) {
+  var object = {};
+  if (nodes) {
+    for (var i = 0, len = nodes.length; i < len; i++) {
+      var node = nodes[i];
+      var attributes = Object.values(getNodeAttributes(node));
+
+      for (var n = 0, len2 = attributes.length; n < len2; n++) {
+        var attribute = attributes[n];
+        object[attribute.name.toLowerCase()] = '';
+      }
+    }
+  }
+  return Object.getOwnPropertyNames(object);
```
Seems like this should also be a Set. (please test)
```suggestion
  if (!nodes) {
    return [];
  }
  var uniqueAttributes = new Set();
  for (var node of nodes) {
    var attributes = Object.values(getNodeAttributes(node));
    for (var attribute of attributes) {
      uniqueAttributes.add(attribute.name.toLowerCase());
    }
  }
  return Array.from(uniqueAttributes);
```

smatei commented on custom_metrics/almanac.js:
```diff
@@ -306,6 +336,57 @@ return JSON.stringify({
+  // Count the mumber of images with srcset and sizes attributes
+  'num_srcset_sizes': document.querySelectorAll('source[srcset][sizes], img[srcset][sizes]').length,
+  // Count the number of imges with srcset with descriptor-x
+  'num_srcset_descriptor_x': document.querySelectorAll('source[srcset*="x"], img[srcset*="x"]').length,
```
I have implemented like this

`'num_srcset_descriptor_x': (() => {

    var nodes = document.querySelectorAll('source[srcset], img[srcset]');

    return Array.from(nodes).filter(node => node.getAttribute('srcset').match(/\s\d+x/)).length;

  })(),`

rviscomi commented:
@smatei thanks for your changes. Reassign back to me when it's ready for review. 👍

smatei commented:
added one more metric for count srcset candidate list size in img only cases (i.e. that are NOT in <picture> ) 

"num_img_not_in_picture_srcset_candidates":2
https://www.webpagetest.org/custom_metrics.php?test=200728_0D_00c8947461e2fdceac3efe1e256dbc78&cached=0

"num_img_not_in_picture_srcset_candidates":0
https://www.webpagetest.org/custom_metrics.php?test=200728_NR_a9794d48c10947f129af889772d35bb3&cached=0

rviscomi commented on custom_metrics/almanac.js:
```diff
@@ -270,7 +270,7 @@ return JSON.stringify({
     };
   })(),
   '09.27': (() => {
-    // Returns a JSON array of nodes with a tabindex and their key/value attributes.
```
Can you revert your changes to this file to prevent merge conflicts?

It's ok to leave `has_picture_img` in there for now.
